File: src/scripted/memoization.nim
## Time-bound memoization for evaluation caching
##
## Provides temporal memoization with timestamp-based cache invalidation.
## Unlike step-based caching (PerAgentCache in ai_core.nim), this uses wall
## clock time which is ideal for:
## - Game tree evaluations that become stale over real time
## - Preventing unbounded memory growth in long sessions
## - Caching expensive computations within frame boundaries
##
## Pattern adopted from metta common/src/metta/common/util/memoization.py

import std/[tables, hashes]
import ../common_types  # for nowSeconds

type
  MemoEntry[V] = object
    value: V
    timestamp: float64

  TimedMemoCache*[K, V] = object
    ## Generic time-bound memoization cache.
    ## Entries expire after maxAge seconds.
    cache: Table[K, MemoEntry[V]]
    maxAge: float64
    lastCleanup: float64
    cleanupInterval: float64  # How often to run cleanup (in seconds)

const
  DefaultMaxAge* = 1.0  ## Default max age for cache entries (1 second)
  DefaultCleanupInterval* = 5.0  ## How often to purge expired entries

proc initTimedMemoCache*[K, V](maxAge: float64 = DefaultMaxAge,
                                cleanupInterval: float64 = DefaultCleanupInterval): TimedMemoCache[K, V] =
  ## Initialize a new time-bound memo cache.
  ## maxAge: entries older than this (in seconds) are considered stale
  ## cleanupInterval: how often to purge expired entries
  result.cache = initTable[K, MemoEntry[V]]()
  result.maxAge = maxAge
  result.lastCleanup = nowSeconds()
  result.cleanupInterval = cleanupInterval

proc cleanup*[K, V](cache: var TimedMemoCache[K, V]) =
  ## Remove all expired entries from the cache.
  ## Called automatically during get() based on cleanupInterval.
  let now = nowSeconds()
  var keysToRemove: seq[K]
  for key, entry in cache.cache.pairs:
    if now - entry.timestamp >= cache.maxAge:
      keysToRemove.add(key)
  for key in keysToRemove:
    cache.cache.del(key)
  cache.lastCleanup = now

proc maybeCleanup[K, V](cache: var TimedMemoCache[K, V], now: float64) {.inline.} =
  ## Run cleanup if enough time has passed since last cleanup.
  if now - cache.lastCleanup >= cache.cleanupInterval:
    cache.cleanup()

proc get*[K, V](cache: var TimedMemoCache[K, V], key: K,
                compute: proc(): V): V =
  ## Get cached value or compute and cache if stale/missing.
  ## The compute proc takes no arguments and returns the result.
  let now = nowSeconds()
  cache.maybeCleanup(now)

  if key in cache.cache:
    let entry = cache.cache[key]
    if now - entry.timestamp < cache.maxAge:
      return entry.value

  # Compute and cache
  result = compute()
  cache.cache[key] = MemoEntry[V](value: result, timestamp: now)

proc getWithArg*[K, V, A](cache: var TimedMemoCache[K, V], key: K, arg: A,
                           compute: proc(a: A): V): V =
  ## Get cached value or compute and cache if stale/missing.
  ## The compute proc takes one argument and returns the result.
  let now = nowSeconds()
  cache.maybeCleanup(now)

  if key in cache.cache:
    let entry = cache.cache[key]
    if now - entry.timestamp < cache.maxAge:
      return entry.value

  # Compute and cache
  result = compute(arg)
  cache.cache[key] = MemoEntry[V](value: result, timestamp: now)

proc getWith2Args*[K, V, A, B](cache: var TimedMemoCache[K, V], key: K, arg1: A, arg2: B,
                               compute: proc(a: A, b: B): V): V =
  ## Get cached value or compute and cache if stale/missing.
  ## The compute proc takes two arguments and returns the result.
  let now = nowSeconds()
  cache.maybeCleanup(now)

  if key in cache.cache:
    let entry = cache.cache[key]
    if now - entry.timestamp < cache.maxAge:
      return entry.value

  # Compute and cache
  result = compute(arg1, arg2)
  cache.cache[key] = MemoEntry[V](value: result, timestamp: now)

proc invalidate*[K, V](cache: var TimedMemoCache[K, V], key: K) =
  ## Manually invalidate a specific cache entry.
  cache.cache.del(key)

proc clear*[K, V](cache: var TimedMemoCache[K, V]) =
  ## Clear all entries from the cache.
  cache.cache.clear()

proc len*[K, V](cache: TimedMemoCache[K, V]): int =
  ## Return the number of entries in the cache (including potentially expired ones).
  cache.cache.len

proc setMaxAge*[K, V](cache: var TimedMemoCache[K, V], maxAge: float64) =
  ## Change the max age for cache entries.
  cache.maxAge = maxAge

# ---------------------------------------------------------------------------
# Convenience: Frame-scoped memoization
# ---------------------------------------------------------------------------
# For computations that should be cached within a single frame but refreshed
# each frame. Uses a very short maxAge (e.g., 0.02 seconds = 50 FPS frame time).

const
  FrameCacheMaxAge* = 0.02  ## ~50 FPS frame time

proc initFrameCache*[K, V](): TimedMemoCache[K, V] =
  ## Initialize a frame-scoped cache (very short max age).
  ## Use for expensive evaluations that don't change within a frame.
  initTimedMemoCache[K, V](maxAge = FrameCacheMaxAge, cleanupInterval = 1.0)

# ---------------------------------------------------------------------------
# Convenience: Position-based cache key helpers
# ---------------------------------------------------------------------------

import vmath

proc posKey*(x, y: int32): int64 {.inline.} =
  ## Create a cache key from a position.
  (int64(x) shl 32) or int64(y)

proc posKey*(pos: IVec2): int64 {.inline.} =
  ## Create a cache key from an IVec2 position.
  posKey(pos.x, pos.y)

proc agentPosKey*(agentId: int, x, y: int32): int64 {.inline.} =
  ## Create a cache key combining agent ID and position.
  ## Uses high bits for agentId, low bits for position hash.
  (int64(agentId) shl 48) or (int64(x) shl 24) or int64(y)

proc agentPosKey*(agentId: int, pos: IVec2): int64 {.inline.} =
  ## Create a cache key combining agent ID and position.
  agentPosKey(agentId, pos.x, pos.y)

# ---------------------------------------------------------------------------
# Convenience: Typed cache key for composite keys
# ---------------------------------------------------------------------------

type
  CacheKey2*[A, B] = object
    ## Composite cache key for two values.
    a*: A
    b*: B

  CacheKey3*[A, B, C] = object
    ## Composite cache key for three values.
    a*: A
    b*: B
    c*: C

proc cacheKey*[A, B](a: A, b: B): CacheKey2[A, B] {.inline.} =
  CacheKey2[A, B](a: a, b: b)

proc cacheKey*[A, B, C](a: A, b: B, c: C): CacheKey3[A, B, C] {.inline.} =
  CacheKey3[A, B, C](a: a, b: b, c: c)

proc hash*[A, B](k: CacheKey2[A, B]): Hash =
  var h: Hash = 0
  h = h !& hash(k.a)
  h = h !& hash(k.b)
  result = !$h

proc hash*[A, B, C](k: CacheKey3[A, B, C]): Hash =
  var h: Hash = 0
  h = h !& hash(k.a)
  h = h !& hash(k.b)
  h = h !& hash(k.c)
  result = !$h

File: tribal_village_env/memoization.py
"""Time-bound memoization for evaluation caching.

Provides temporal memoization with timestamp-based cache invalidation.
Unlike step-based caching, this uses wall clock time which is ideal for:
- Game tree evaluations that become stale over real time
- Preventing unbounded memory growth in long sessions
- Caching expensive computations within frame boundaries

Pattern adopted from metta common/src/metta/common/util/memoization.py
"""

from __future__ import annotations

import time
from collections.abc import Awaitable, Callable, Hashable
from functools import wraps
from typing import Any, Generic, ParamSpec, TypeVar

P = ParamSpec("P")
T = TypeVar("T")
K = TypeVar("K", bound=Hashable)
V = TypeVar("V")

# Default cache parameters
DEFAULT_MAX_AGE = 1.0  # Default max age for cache entries (1 second)
DEFAULT_CLEANUP_INTERVAL = 5.0  # How often to purge expired entries
FRAME_CACHE_MAX_AGE = 0.02  # ~50 FPS frame time


def memoize(max_age: float) -> Callable[[Callable[P, Awaitable[T]]], Callable[P, Awaitable[T]]]:
    """Async time-bound memoization decorator.

    Caches function results with timestamp-based cache invalidation.
    Entries older than max_age seconds are considered stale.

    Args:
        max_age: Maximum age in seconds before cache entries expire.

    Returns:
        Decorator that wraps async functions with time-bound caching.

    Example:
        @memoize(max_age=1.0)
        async def expensive_evaluation(state_hash: int) -> float:
            # ... expensive computation ...
            return result
    """

    def decorator(func: Callable[P, Awaitable[T]]) -> Callable[P, Awaitable[T]]:
        cache: dict[tuple[Any, ...], tuple[T, float]] = {}

        @wraps(func)
        async def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
            key = (args, tuple(sorted(kwargs.items())))
            current_time = time.time()

            if key in cache:
                value, timestamp = cache[key]
                if current_time - timestamp < max_age:
                    return value

            result = await func(*args, **kwargs)
            cache[key] = (result, current_time)
            return result

        return wrapper

    return decorator


def memoize_sync(max_age: float) -> Callable[[Callable[P, T]], Callable[P, T]]:
    """Synchronous time-bound memoization decorator.

    Same as memoize() but for synchronous functions.

    Args:
        max_age: Maximum age in seconds before cache entries expire.

    Returns:
        Decorator that wraps sync functions with time-bound caching.

    Example:
        @memoize_sync(max_age=0.5)
        def score_enemy(agent_id: int, enemy_id: int) -> float:
            # ... expensive scoring ...
            return score
    """

    def decorator(func: Callable[P, T]) -> Callable[P, T]:
        cache: dict[tuple[Any, ...], tuple[T, float]] = {}

        @wraps(func)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
            key = (args, tuple(sorted(kwargs.items())))
            current_time = time.time()

            if key in cache:
                value, timestamp = cache[key]
                if current_time - timestamp < max_age:
                    return value

            result = func(*args, **kwargs)
            cache[key] = (result, current_time)
            return result

        return wrapper

    return decorator


class TimedMemoCache(Generic[K, V]):
    """Generic time-bound memoization cache.

    Entries expire after max_age seconds. Provides explicit cache management
    for cases where decorator-based memoization isn't suitable.

    This mirrors the Nim implementation in src/scripted/memoization.nim.

    Args:
        max_age: Maximum age in seconds before cache entries expire.
        cleanup_interval: How often to purge expired entries (in seconds).

    Example:
        cache: TimedMemoCache[int, float] = TimedMemoCache(max_age=1.0)

        def get_evaluation(state_hash: int) -> float:
            return cache.get(state_hash, lambda: expensive_compute(state_hash))
    """

    def __init__(
        self,
        max_age: float = DEFAULT_MAX_AGE,
        cleanup_interval: float = DEFAULT_CLEANUP_INTERVAL,
    ) -> None:
        self._cache: dict[K, tuple[V, float]] = {}
        self._max_age = max_age
        self._cleanup_interval = cleanup_interval
        self._last_cleanup = time.time()

    @property
    def max_age(self) -> float:
        """Maximum age for cache entries in seconds."""
        return self._max_age

    @max_age.setter
    def max_age(self, value: float) -> None:
        """Set the maximum age for cache entries."""
        self._max_age = value

    def cleanup(self) -> None:
        """Remove all expired entries from the cache."""
        now = time.time()
        expired_keys = [
            key
            for key, (_, timestamp) in self._cache.items()
            if now - timestamp >= self._max_age
        ]
        for key in expired_keys:
            del self._cache[key]
        self._last_cleanup = now

    def _maybe_cleanup(self, now: float) -> None:
        """Run cleanup if enough time has passed since last cleanup."""
        if now - self._last_cleanup >= self._cleanup_interval:
            self.cleanup()

    def get(self, key: K, compute: Callable[[], V]) -> V:
        """Get cached value or compute and cache if stale/missing.

        Args:
            key: Cache key.
            compute: Zero-argument callable to compute the value if needed.

        Returns:
            Cached or freshly computed value.
        """
        now = time.time()
        self._maybe_cleanup(now)

        if key in self._cache:
            value, timestamp = self._cache[key]
            if now - timestamp < self._max_age:
                return value

        result = compute()
        self._cache[key] = (result, now)
        return result

    def get_with_arg(self, key: K, arg: Any, compute: Callable[[Any], V]) -> V:
        """Get cached value or compute with one argument.

        Args:
            key: Cache key.
            arg: Argument to pass to compute function.
            compute: Single-argument callable to compute the value.

        Returns:
            Cached or freshly computed value.
        """
        now = time.time()
        self._maybe_cleanup(now)

        if key in self._cache:
            value, timestamp = self._cache[key]
            if now - timestamp < self._max_age:
                return value

        result = compute(arg)
        self._cache[key] = (result, now)
        return result

    def get_with_2_args(
        self, key: K, arg1: Any, arg2: Any, compute: Callable[[Any, Any], V]
    ) -> V:
        """Get cached value or compute with two arguments.

        Args:
            key: Cache key.
            arg1: First argument to pass to compute function.
            arg2: Second argument to pass to compute function.
            compute: Two-argument callable to compute the value.

        Returns:
            Cached or freshly computed value.
        """
        now = time.time()
        self._maybe_cleanup(now)

        if key in self._cache:
            value, timestamp = self._cache[key]
            if now - timestamp < self._max_age:
                return value

        result = compute(arg1, arg2)
        self._cache[key] = (result, now)
        return result

    def invalidate(self, key: K) -> None:
        """Manually invalidate a specific cache entry."""
        self._cache.pop(key, None)

    def clear(self) -> None:
        """Clear all entries from the cache."""
        self._cache.clear()

    def __len__(self) -> int:
        """Return the number of entries (including potentially expired ones)."""
        return len(self._cache)

    def __contains__(self, key: K) -> bool:
        """Check if key exists and is not expired."""
        if key not in self._cache:
            return False
        _, timestamp = self._cache[key]
        return time.time() - timestamp < self._max_age


def create_frame_cache() -> TimedMemoCache[Any, Any]:
    """Create a frame-scoped cache (very short max age).

    Use for expensive evaluations that don't change within a frame.
    Max age is approximately one frame at 50 FPS.

    Returns:
        TimedMemoCache configured for frame-scoped caching.
    """
    return TimedMemoCache(max_age=FRAME_CACHE_MAX_AGE, cleanup_interval=1.0)


# ---------------------------------------------------------------------------
# Convenience: Cache key helpers
# ---------------------------------------------------------------------------


def pos_key(x: int, y: int) -> int:
    """Create a cache key from a position.

    Args:
        x: X coordinate.
        y: Y coordinate.

    Returns:
        Integer cache key combining both coordinates.
    """
    return (x << 32) | (y & 0xFFFFFFFF)


def agent_pos_key(agent_id: int, x: int, y: int) -> int:
    """Create a cache key combining agent ID and position.

    Uses high bits for agentId, low bits for position hash.

    Args:
        agent_id: Agent identifier.
        x: X coordinate.
        y: Y coordinate.

    Returns:
        Integer cache key combining agent and position.
    """
    return (agent_id << 48) | (x << 24) | (y & 0xFFFFFF)


def cache_key(*args: Hashable) -> tuple[Hashable, ...]:
    """Create a composite cache key from multiple values.

    Args:
        *args: Hashable values to combine into a key.

    Returns:
        Tuple suitable for use as a dictionary key.
    """
    return args
